src/com/google/cose/Ec2SigningKey.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cose;
+
+import co.nstant.in.cbor.CborException;
+import co.nstant.in.cbor.model.Array;
+import co.nstant.in.cbor.model.ByteString;
+import co.nstant.in.cbor.model.DataItem;
+import co.nstant.in.cbor.model.Map;
+import co.nstant.in.cbor.model.NegativeInteger;
+import co.nstant.in.cbor.model.UnsignedInteger;
+import com.google.cose.exceptions.CoseException;
+import com.google.cose.utils.Algorithm;
+import com.google.cose.utils.CborUtils;
+import com.google.cose.utils.CoseUtils;
+import com.google.cose.utils.Headers;
+import java.math.BigInteger;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** Implements EC2 COSE_Key spec for signing purposes. */
+public final class Ec2SigningKey extends CoseKey {
+  private static final int SIGN_POSITIVE = 1;
+
+  KeyPair keyPair;
+
+  public Ec2SigningKey(DataItem cborKey) throws CoseException, CborException {
+    super(cborKey);
+
+    keyPair = getKeyPairFromCbor();
+    if ((operations == null)
+        || (operations.contains(Headers.KEY_OPERATIONS_VERIFY)
+        && operations.contains(Headers.KEY_OPERATIONS_SIGN))) {
+      return;
+    }
+    throw new CoseException("Signing key requires sign and verify operations.");
+  }
+
+  static class Builder {
+    private String keyId;
+    private Algorithm algorithm;
+    private final List<Integer> operations;
+    private byte[] baseIv;
+    private Integer curve = null;
+    private byte[] xCor;
+    private byte[] yCor;
+    private byte[] dCor;
+
+    Builder() {
+      operations = new ArrayList<>();
+    }
+
+    public Ec2SigningKey build() throws CoseException, CborException {
+      if ((curve == null) || (dCor == null && (xCor == null || yCor == null))) {
+        throw new CoseException("Need curve and key material information.");
+      }
+      if (xCor == null ^ yCor == null) {
+        // If we have only one public key coordinate, raise an exception
+        throw new CoseException("Need both x and y coordinate information for EC2 public key.");
+      }
+      if (operations.size() != 0 && !operations.contains(Headers.KEY_OPERATIONS_VERIFY)
+          && !operations.contains(Headers.KEY_OPERATIONS_SIGN)) {
+        throw new CoseException("Need Sign and Verify operation for the signing key.");
```

Comment:
Nit: `Sign or Verify`

Reply:
we can reach this a user constructs a CBOR key separately and provides it for Key construction.

Reply:
That would hold true if user only calls withOperations only once. However, if user makes multiple calls to withOperations to add different operations, we may mistakenly throw an exception where none was expected.

Reply:
But `withOperations` should only allow KEY_OPERATIONS_VERIFY or KEY_OPERATIONS_SIGN. So it doesn't matter how many times `withOperations` is called right? It's impossible to build a key with invalid operations so long as the builder does its job. Maybe I'm missing something?

Reply:
Nah. I missed the context on allowing only 2 operations for each kind of key. It does make sense to allow only these specific operations on the key since we do not have any other cryptographic operation in the key anyways.

We also restrict the usage of the encoded cose key blobs to only one function this way.